packages/@ama-sdk/schematics/schematics/typescript/core/helpers/copy-referenced-files.ts

```diff
@@ -0,0 +1,96 @@
+import { existsSync } from 'node:fs';
+import { copyFile, mkdir, readFile, rm } from 'node:fs/promises';
+import { dirname, join, normalize, posix, relative, resolve } from 'node:path';
+
+const refMatcher = /\B['"]?[$]ref['"]?\s*:\s*([^#\n]+)/g;
+
+/**
+ * Extract the list of local references from a single spec file content
+ * @param specContent
+ * @param basePath
+ */
+function extractRefPaths(specContent: string, basePath: string): string[] {
+  const refs = specContent.match(refMatcher);
+  return refs ?
+    refs
+      .map((capture) => capture.replace(refMatcher, '$1').replace(/['"]/g, ''))
+      .filter((refPath) => refPath.startsWith('.'))
+      .map((refPath) => join(basePath, refPath))
+    : [];
+}
+
+/**
+ * Recursively extract the list of local references starting from the input spec file
+ * @param specFilePath
+ * @param referenceFilePath
+ * @param visited
+ */
+async function extractRefPathRecursive(specFilePath: string, referenceFilePath: string, visited: Set<string>): Promise<string[]> {
+  const resolvedFilePath = resolve(specFilePath);
+  if (!visited.has(resolvedFilePath)) {
+    visited.add(resolvedFilePath);
+
+    const specContent = await readFile(specFilePath, {encoding: 'utf8'});
+    const refPaths = extractRefPaths(specContent, relative(dirname(referenceFilePath), dirname(specFilePath)));
+    const recursiveRefPaths = await Promise.all(
+      refPaths.map((refPath) => extractRefPathRecursive(join(dirname(referenceFilePath), refPath), referenceFilePath, visited))
+    );
+    return [
+      ...refPaths,
+      ...recursiveRefPaths.flat()
+    ];
+  }
+  return [];
+}
+
+/**
+ * Replace all the local relative references using the new base relative path
+ * @param specContent
+ * @param newBaseRelativePath
+ */
+export function updateLocalRelativeRefs(specContent: string, newBaseRelativePath: string) {
+  const formatPath = (inputPath:string) => (inputPath.startsWith('.') ? inputPath : `./${inputPath}`).replace(/\\+/g, '/');
+  return specContent.replace(refMatcher, (match, ref: string) => {
+    const refPath = ref.replace(/['"]/g, '');
+    return refPath.startsWith('.') ?
+      match.replace(refPath, formatPath(normalize(posix.join(newBaseRelativePath, refPath))))
+      : match;
+  });
+}
+
+/**
+ * Copy the local files referenced in the input spec file to the output directory
+ * @param specFilePath
+ * @param outputDirectory
+ */
+export async function copyReferencedFiles(specFilePath: string, outputDirectory: string) {
+  const dedupe = (paths: string[]) =>
```

Comment:
For each string from paths, the findIndex method will return the first element in the array to match the comparison, but wouldn't one iteration be enough here to dedupe the array ?

Reply:
You could use maps to avoid the double iteration

Reply:
`const dedupe = (paths: string[]) => ([...new Set(paths)]);` shouldn't be enough ?

Reply:
some path are equivalent but not equal as strings 
say you have
```
dirA
|  spec.yml
|  ref.yml
```
the path to ref.yml can be `./ref.yml` or `../dirA/ref.yml`, that's why I was not just using a Set